tools/universe/s3_uploader.py

```diff
@@ -36,6 +38,55 @@ def upload(self, filepath, content_type=None):
         cmdlist.append('{} {} 1>&2'.format(filepath, dest_url))
         cmd = ' '.join(cmdlist)
         log.info(cmd)
+
+        # try once:
         ret = os.system(cmd)
-        if not ret == 0:
+        if ret != 0:
+            if self._renew_credentials():
+                # renew succeeded, try again:
+                ret = os.system(cmd)
+        if ret != 0:
+            # failed once, then renewal failed or retry failed
             raise Exception('Failed to upload {} to {}'.format(filepath, dest_url))
+
+    def _renew_credentials(self):
+        if self._reauth_attempted:
+            # reauth was already tried once during this session. any failure must be from something else.
+            return False
+        self._reauth_attempted = True
+
+        # check that maws is installed before trying anything else
+        if subprocess.run('maws -h'.split(), stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL).returncode != 0:
+            log.error('Unable to renew credentials: Missing "maws" command in PATH')
+            return False
+
+        # renew creds using discovered aws profile
+        aws_profile = self._get_aws_profile()
+        log.info('Upload attempt failed, renewing credentials for AWS profile {}'.format(aws_profile))
+        ret = os.system('maws login {}'.format(aws_profile))
```

Comment:
Newer versions of `maws` have a `-r` flag which automatically refreshes the credentials. This would have to be started in a separate thread.

Reply:
In theory, this path should only be hit if the user hasn't already set that up